documentation/jetty-documentation/src/main/asciidoc/programming-guide/arch-threads.adoc

```diff
@@ -0,0 +1,200 @@
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+[[pg-arch-threads]]
+=== Jetty Threading Architecture
+
+Writing a performant client or server is difficult, because it should:
+
+* Scale well with the number of processors.
+* Be efficient at using processor caches to avoid link:https://en.wikipedia.org/wiki/Parallel_slowdown[parallel slowdown].
+* Support multiple network protocols that may have very different requirements; for example, multiplexed protocols such as HTTP/2 introduce new challenges that are not present in non-multiplexed protocols such as HTTP/1.1.
+* Support different application threading models; for example, if a Jetty server invokes server-side application code that is allowed to call blocking APIs, then the Jetty server should not be affected by how long the blocking API call is, and should be able to process other connections or other requests in a timely fashion.
+
+[[pg-arch-threads-execution-strategy]]
+==== Execution Strategies
+
+The Jetty threading architecture can be modeled with a producer/consumer pattern, where produced tasks needs to be consumed efficiently.
+
+For example, Jetty produces (among others) these tasks:
+
+* A task that wraps a NIO selection event, see the xref:pg-arch-io[Jetty I/O architecture].
+* A task that wraps the invocation of application code that may block (for example, the invocation of a Servlet to handle an HTTP request).
+
+A task is typically a `Runnable` object that may implement `org.eclipse.jetty.util.thread.Invocable` to indicate the behavior of the task (in particular, whether the task may block or not).
+
+Once a task has been produced, it may be consumed using these strategies:
+
+* `Produce-Consume`
+* `Produce-Execute-Consume`
+* `Execute-Produce-Consume`
+
+[[pg-arch-threads-execution-strategy-pc]]
+===== Produce-Consume
+In the `Produce-Consume` mode, the producer thread loops to produce a task that is run directly by the `Producer Thread`.
+
+[plantuml]
+----
+skinparam backgroundColor transparent
+
+compact concise "Producer Thread" as PT
+hide time-axis
+
+@PT
+0 is T1 #lightgreen
+1 is "Run T1" #dodgerblue
+5 is T2 #lightgreen
+6 is "Run T2" #dodgerblue
+8 is T3 #lightgreen
+9 is "Run T3" #dodgerblue
+12 is {hidden}
+----
+
+If the task is a NIO selection event, then this model is the thread-per-selector model which is very CPU core cache efficient, but suffers from the link:http://en.wikipedia.org/wiki/Head-of-line_blocking[head-of-line blocking]: if one of the tasks blocks or runs slowly, then subsequent tasks cannot be produced (and therefore cannot be consumed either) and will pay in latency the cost of running previous, possibly unrelated, tasks.
```

Comment:
Maybe we should include hints about the Adaptive Strategy:
```suggestion
If the task is a NIO selection event, then this model is the thread-per-selector model which is very CPU core cache efficient, but suffers from the link:http://en.wikipedia.org/wiki/Head-of-line_blocking[head-of-line blocking]: if one of the tasks blocks or runs slowly, then subsequent tasks cannot be produced (and therefore cannot be consumed either) and will pay in latency the cost of running previous, possibly unrelated, tasks.

This strategy should only be used if the produced task is known not to block or the deployment is able to suffer head of line blocking.
```

Reply:
Slightly modified, but done.